tests/SchedulerTest.php

```diff
@@ -0,0 +1,221 @@
+        $promisesCount = [];
+        $task = new PromiseBoundTask((new Promise\Deferred())->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_CANCEL, function (Task $_, array $promises) use (&$promisesCount) {
+                $promisesCount = $promises;
+            });
+
+        // We have to use addtimer!!
+        Loop::addTimer(0, function () use ($task) {
+            $this->scheduler->remove($task);
+        });
+
+        $this->runOff();
+
+        $this->assertCount(2, $promisesCount, 'Scheduler::remove() could not cancel running tasks');
+    }
+
+    public function testFailedTasks()
+    {
+        $taskFailed = false;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use ($deferred) {
+                $deferred->reject(new TaskRejectedException('Exited mysteriously!'));
+            })
+            ->on(CountableScheduler::ON_TASK_FAILED, function (Task $t, Throwable $_) use (&$taskFailed) {
+                $taskFailed = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($taskFailed, 'Scheduler could not handle failed tasks');
+    }
+
+    public function testTaskRunSuccess()
+    {
+        $taskFulFilled = false;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use ($deferred) {
+                $deferred->resolve(10);
+            })
+            ->on(CountableScheduler::ON_TASK_DONE, function (Task $_, $any) use (&$taskFulFilled) {
+                $taskFulFilled = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($taskFulFilled, 'Scheduler could not run task successfully');
+    }
+
+    public function testPropagateTaskExceptions()
+    {
+        $this->expectException(TaskRejectedException::class);
+
+        $exception = null;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $t, ExtendedPromiseInterface $_) use ($deferred) {
+                $deferred->reject(new TaskRejectedException('Rejected!'));
+            })
+            ->on(CountableScheduler::ON_TASK_FAILED, function (Task $_, Throwable $reason) use (&$exception) {
+                $exception = $reason;
+            });
+
+        $this->runOff();
+
+        throw $exception;
+    }
+
+    public function testSpawningNewTasks()
```

Comment:
This is basically just a test for count. I would test all counts and rename it to `testCountsWithMultipleScheduledTasks()`.